Improve POSIX-compliance of fmemopen

arichardson opened the pull request:
I was trying to run a program that uses fmemopen with picolibc and got unexpected EOF returns when reading a buffer.
It turned out picolibc was returning EOF on the first NUL character which is not compliant with POSIX.

After fixing this initial behaviour difference I tried running the NetBSD fmemopen  test to check compliance and fixed various other issues found while running that test. The test assumes some NetBSD-speicific behaviour that I did not implement here, so I lightly modified it with a few ifdefs.

keith-packard commented on newlib/libc/tinystdio/fmemopen.c:
```diff
@@ -33,119 +33,166 @@
+        pos += mf->size;
+        break;
+    }
+    if (pos < 0 || mf->bufsize < (size_t)pos)
```
Hrm. I think we need to be careful here with types -- if 'off_t' is larger than 'size_t', then you could pass a large
'pos' which truncated to an in-range value. It's "probably" safe to assume `sizeof(off_t) >= sizeof(size_t)`?

keith-packard reviewed (changes requested):
Looks good so far; I think the seek function needs some careful analysis of off_t vs size_t usage.

arichardson commented:
> Looks good so far; I think the seek function needs some careful analysis of off_t vs size_t usage.

I've added a static_assert and changed the check in seek to use off_t.

arichardson commented:
Native tests should also be passing now. I also now ensure NUL byte insertion if there is space matching the comments on https://austingroupbugs.net/view.php?id=657#c6535

arichardson commented on test/t_fmemopen.c:
```diff
@@ -30,26 +30,35 @@
+#define _GNU_SOURCE  /* strnlen */
+#elif defined(__linux__)
 #define _GNU_SOURCE
 #include <features.h>
```
Somewhat unrelated comment: I initially just added a __PICOLIBC__ case as follows:
```
#if defined(__linux__)
#define _GNU_SOURCE
#include <features.h>
#endif
#if defined(__PICOLIBC__)
#define _GNU_SOURCE  /* strnlen */
#endif 
```

Interestingly that broke the tests since including features.h defines __GLIBC__ and then the wrong branches are compiled. I wonder if picolibc should install an empty features.h to avoid this subtle bug?

keith-packard replied:
Hrm. We've got sys/features.h which appears to do something similar. Maybe a top-level features.h that includes sys/features.h (or picolibc.h?)

arichardson replied:
Actually, maybe the real solution is to remove system includes from the search path using --sysroot? But anyway this is unrelated and should IMO not block this PR.

keith-packard replied:
sorry, been afk for the weekend. thanks much for this fix.
